controllers/component_controller_test.go

```diff
@@ -0,0 +1,562 @@
+/*
+Copyright 2021 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	data "github.com/devfile/library/pkg/devfile/parser/data"
+	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	devfile "github.com/redhat-appstudio/application-service/pkg/devfile"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	//+kubebuilder:scaffold:imports
+)
+
+var _ = Describe("Component controller", func() {
+
+	// Define utility constants for object names and testing timeouts/durations and intervals.
+	const (
+		HASAppName      = "test-hasapplication-123"
+		HASCompName     = "test-hascomponent-123"
+		HASAppNamespace = "default"
+		DisplayName     = "petclinic"
+		Description     = "Simple petclinic app"
+		ComponentName   = "backend"
+		SampleRepoLink  = "https://github.com/devfile-samples/devfile-sample-java-springboot-basic"
+	)
+
+	Context("Create Component with basic field set", func() {
```

Comment:
Could you add a simple case here where the devfile is malformed? Maybe by pointing `devfileURL` to a devfile that cannot be parsed into a valid devfile?

Reply:
And maybe another case where the `Application` CR referenced doesn't exist?